File: kotlin/874-backspace-string-compare-Easy/Solution.kt
class Solution {
    fun backspaceCompare(s: String, t: String): Boolean {
        return norm(s) == norm(t)
    }
    
    private fun norm(s: String): String {
        val b = StringBuilder()
        var skip = 0
        for (i in s.lastIndex downTo 0) {
            when {
                s[i] == '#' -> skip++
                skip > 0 -> skip--
                else -> b.append(s[i])
            }
        }
        return b.toString()
    }
}

File: scala/874-backspace-string-compare-Easy/Solution.scala
object Solution {
    def backspaceCompare(S: String, T: String): Boolean = {
        edit(S) == edit(T)
    }
    private def edit(s: String): String = {
        var b = new StringBuilder()
        var skip = 0
        for (i <- s.length - 1 to 0 by -1) {
            if (s(i) == '#')
                skip += 1
            else if (skip > 0)
                skip -= 1
            else
                b += s(i)     
        }
        b.reverse.result
    }
}